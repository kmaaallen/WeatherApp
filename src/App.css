.App {
  text-align: center;
  max-height: 100vh;
}

.Background {
  background-size: cover;
  animation: 10s fadeIn;
  height: 100vh;
  z-index: 25;
  position: relative;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}