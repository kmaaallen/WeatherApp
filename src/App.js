import './App.css';
import React, { useEffect, useLayoutEffect, useState, useRef } from 'react';
//Components
import SearchWeather from './components/SearchWeather';
import CurrentWeather from './components/CurrentWeather';
import Background from './components/Background';
//Backgrounds
import Atmosphere from './images/Atmosphere.jpg';
import Clouds from './images/Clouds.jpg';
import Rain from './images/Rain.jpg';
import Snow from './images/Snow.jpg';
import Clear from './images/Clear.jpg';
import Thunderstorm from './images/Thunderstorm.jpg';
import Drizzle from './images/Drizzle.jpg';

//TODO: Loading UI
//TODO: Error catching
//TODO: transition background smoother

function App() {
  const [latitude, setLatitude] = useState('');
  const [longitude, setLongitude] = useState('');
  const [data, setData] = useState([]);
  const [background, setBackground] = useState([]);

  const firstCall = useRef(true);

  function myCallBack(cityData) {
    setData(cityData);
  };

  useLayoutEffect(() => {
    if (data.length === 0) {
      return;
    } else {
      const atmosphere = ['Mist', 'Smoke', 'Haze', 'Dust', 'Fog', 'Sand', 'Ash', 'Squall', 'Tornado'];
      if (atmosphere.includes(data.weather[0].main)) setBackground(Atmosphere);
      if (data.weather[0].main === 'Clouds') setBackground(Clouds);
      if (data.weather[0].main === 'Rain') setBackground(Rain);
      if (data.weather[0].main === 'Snow') setBackground(Snow);
      if (data.weather[0].main === 'Drizzle') setBackground(Drizzle);
      if (data.weather[0].main === 'Thunderstorm') setBackground(Thunderstorm);
      if (data.weather[0].main === 'Clear') setBackground(Clear);
    }
  }, [data])

  useEffect(() => {
    navigator.geolocation.getCurrentPosition(function (pos) {
      setLatitude(pos.coords.latitude);
      setLongitude(pos.coords.longitude);
    });
  }, []);

  useLayoutEffect(() => {
    if (firstCall.current) {
      firstCall.current = false;
      return;
    } else if (latitude !== '' && longitude !== '') {
      fetch(`/api/weather/${latitude}/${longitude}`)
        .then((res) => res.json())
        .then((data) => {
          setData(data.data);
        });
    }
  }, [latitude, longitude]);

  return (
    //style={{ backgroundImage: `url(${background})` }}
    <div className="App" >
      <Background background={background} >
        <SearchWeather callBackFromParent={myCallBack} />
        <CurrentWeather data={data} />
      </Background>
    </div >
  );
}

export default App;
