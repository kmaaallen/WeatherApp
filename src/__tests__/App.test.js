import React from 'react';
import renderer from 'react-test-renderer';
import { render, waitForElementToBeRemoved } from '@testing-library/react';
import App from '../App';

describe('<App />', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<App />).toJSON();
        expect(tree).toMatchSnapshot();

        const { getByText } = render(<App />);
        const message = getByText('Please share your location to see local weather, or use the search bar');
        expect(message).toBeTruthy();
    });
    it('should show local weather on load from users coordinates', () => {
        const mockGeolocation = {
            getCurrentPosition: jest.fn()
                .mockImplementationOnce((success) => Promise.resolve(success({
                    coords: {
                        latitude: 37.39,
                        longitude: -122.08
                    }
                })))
        };
        global.navigator.geolocation = mockGeolocation;
        const { findByText, queryByText } = render(<App />);
        waitForElementToBeRemoved(queryByText('Loading')).then(() =>
            expect(findByText('Mountain View, US')).toBeTruthy(),
        )
    });
});
