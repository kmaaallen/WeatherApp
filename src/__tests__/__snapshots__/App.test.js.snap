// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="Background"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  />
  <form
    className="makeStyles-form-1"
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-search-2"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
        data-shrink={false}
        htmlFor="city-text"
        id="city-text-label"
      >
        City
      </label>
      <div
        className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiFilledInput-input"
          disabled={false}
          id="city-text"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
      
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-3"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Search
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </form>
  <div
    className="makeStyles-container-4"
  >
    <div>
      <div>
        Loading
      </div>
      <div
        className="MuiCircularProgress-root makeStyles-progressCircle-11 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
